pkg/federate/kubefed/resource_test.go

```diff
@@ -0,0 +1,389 @@
+		It("should throw an error", func() {
+			err := f.Distribute(resource, clusterNames...)
+			Expect(err).To(HaveOccurred())
+		})
+	})
+
+	When("the resource type scheme has been added to the type registry", func() {
+		BeforeEach(func() {
+			err := corev1.AddToScheme(scheme)
+			Expect(err).ToNot(HaveOccurred())
+			fedResource, _ = createFederatedResource(scheme, resource)
+		})
+
+		It("should return no error on success", func() {
+			err := f.Distribute(resource, clusterNames...)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		When("the resource is already in the Kube API", func() {
+			expectedImage := "apache"
+
+			BeforeEach(func() {
+				initObjs = append(initObjs, fedResource)
+				resource = newPod("test-pod", expectedImage)
+			})
+
+			It("should update the resource", func() {
+				_ = f.Distribute(resource, clusterNames...)
+				fedPod, err := getResourceFromAPI(f.kubeFedClient, fedResource)
+				Expect(err).ToNot(HaveOccurred())
+				fedPodObj, _, err := getTemplateField(fedPod)
+				Expect(err).ToNot(HaveOccurred())
+				expectObj, _, err := getTemplateField(fedResource)
+				Expect(err).ToNot(HaveOccurred())
+
+				Expect(fedPodObj["spec"]).ToNot(BeEquivalentTo(expectObj["spec"]))
+
+				mapObj, _, _ := unstructured.NestedSlice(
+					fedPod.Object,
+					ctlutil.SpecField, ctlutil.TemplateField, "spec", "containers",
+				)
+				image := mapObj[0].(map[string]interface{})["image"]
+				Expect(image).To(BeEquivalentTo(expectedImage))
+			})
+		})
+
+		When("on update the Kube API returns an error that is not a NotFound", func() {
+			BeforeEach(func() {
+				clientPatcher = &fakeClientWithUpdateError{"test error"}
+			})
+
+			It("should return an error and not create the resource", func() {
+				err := f.Distribute(resource, clusterNames...)
+				Expect(err).To(MatchError(HavePrefix("test error")))
+			})
+		})
+
+		When("the resource is not in the Kube API", func() {
+			It("should create the resource", func() {
+				// Sanity check
+				_, err := getResourceFromAPI(f.kubeFedClient, fedResource)
+				Expect(err).To(HaveOccurred())
+
+				_ = f.Distribute(resource, clusterNames...)
+				fedPod, err := getResourceFromAPI(f.kubeFedClient, fedResource)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(fedPod).ToNot(BeNil())
+				Expect(fedPod).To(BeEquivalentTo(fedResource))
+			})
+		})
+
+		When("the resource needs to be distributed to specific clusters", func() {
+			BeforeEach(func() {
+				clusterNames = []string{"cluster1", "cluster2"}
+			})
+
+			It("should set the placement to be an explicit list of clusters", func() {
+				// This test implicitly tests that the nested hierarchy of fields exists
+				_ = f.Distribute(resource, clusterNames...)
+				fedPod, err := getResourceFromAPI(f.kubeFedClient, fedResource)
+				Expect(err).ToNot(HaveOccurred())
+
+				mapObj, found, err := getClustersField(fedPod)
+				fedClusterNames := unpackClusterNames(mapObj)
```

Comment:
bit of a nit but this line should really be after 279 as it technically shouldn't try to use `mapObj` if an error or `!found` occurred. If `mapObj` is nil, `unpackClusterNames` does happen to handle that and not blow up.

Reply:
fair enough